sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,492 @@
+ *
+ * <pre>{@code
+ *
+ * pipeline.apply(
+ *   MqttIO.read()
+ *    .withMqttConnectionConfiguration(MqttIO.MqttConnectionConfiguration.create(
+ *      "tcp://host:11883",
+ *      "my_client_id",
+ *      "my_topic"))
+ *
+ * }</pre>
+ *
+ * <h3>Writing to a MQTT broker</h3>
+ *
+ * <p>MqttIO sink supports writing {@code byte[]} to a topic on a MQTT broker.</p>
+ *
+ * <p>To configure a MQTT sink, as for the read, you have to specify a MQTT connection
+ * configuration with {@code ClientId}, {@code ServerURI}, {@code Topic}.</p>
+ *
+ * <p>Eventually, you can also specify the {@code Retained} and {@code QoS} of the MQTT
+ * message.</p>
+ *
+ * <p>For instance:</p>
+ *
+ * <pre>{@code
+ *
+ * pipeline
+ *   .apply(...) // provide PCollection<byte[]>
+ *   .MqttIO.write()
+ *     .withMqttConnectionConfiguration(MqttIO.MqttConnectionConfiguration.create(
+ *       "tcp://host:11883",
+ *       "my_client_id",
+ *       "my_topic"))
+ *     .withRetained(true)
+ *     .withQoS(2)
+ *
+ * }</pre>
+ */
+public class MqttIO {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MqttIO.class);
+
+  public static Read read() {
+    return new AutoValue_MqttIO_Read.Builder()
+        .setMaxReadTime(null).setMaxNumRecords(Long.MAX_VALUE).build();
+  }
+
+  public static Write write() {
+    return new AutoValue_MqttIO_Write.Builder()
+        .setRetained(false)
+        .setQos(0)
+        .build();
+  }
+
+  private MqttIO() {
+  }
+
+  /**
+   * A POJO describing a MQTT connection.
+   */
+  @AutoValue
+  abstract static class MqttConnectionConfiguration implements Serializable {
+
+    @Nullable abstract String serverUri();
+    @Nullable abstract String clientId();
+    @Nullable abstract String topic();
+
+    public static MqttConnectionConfiguration create(String serverUri, String clientId) {
+      checkNotNull(serverUri, "serverUri");
+      checkNotNull(clientId, "clientId");
+      return new AutoValue_MqttIO_MqttConnectionConfiguration(serverUri, clientId, null);
+    }
+
+    public static MqttConnectionConfiguration create(String serverUri, String clientId,
+                                                     String topic) {
+      checkNotNull(serverUri, "serverUri");
+      checkNotNull(clientId, "clientId");
+      return new AutoValue_MqttIO_MqttConnectionConfiguration(serverUri, clientId, topic);
+    }
+
+    private void validate() {
+      checkNotNull(serverUri());
+      checkNotNull(clientId());
+    }
+
+    private void populateDisplayData(DisplayData.Builder builder) {
+      builder.add(DisplayData.item("serverUri", serverUri()));
+      builder.add(DisplayData.item("clientId", clientId()));
+      builder.add(DisplayData.item("topic", topic()));
+    }
+
+    MqttClient getClient() throws Exception {
+      MqttClient client = new MqttClient(serverUri(), clientId());
+      client.connect();
+      if (topic() != null) {
+        client.subscribe(topic());
+      }
```

Comment:
If topic is null, then we don't subscribe at all? What does it mean to have a null topic?